### PR TITLE
[Concept Exercise] HighScore -- Hashes

### PR DESCRIPTION
**Pull request by neenjaw**
Basically a direct translation of the high-score exercise in Elixir

WIP

### CODE REVIEW

**Comment by neenjaw** [comment]
Forgive the ugly label, not all tracks get pretty ones I guess.

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ bin/configlet.exe
 coverage/
 .last_generator_run
 node_modules/
+.tool-versions
```
Comment (kotp) [inline .gitignore:7]: This doesn't seem to be relevant to the exercise.

File: exercises/concept/high-score/.meta/exemplar.rb [hunk 2]
```diff
@@ -1,31 +1,31 @@
...
-    score_board
+  def add_player(name, initial_score = INITIAL_SCORE)
+    @scores[name] = initial_score
   end
 
-  def self.remove_player(score_board, name)
-    score_board.delete(name)
-    score_board
+  def remove_player(name)
+    @scores.delete(name)
```
Comment (kotp) [inline exercises/concept/high-score/.meta/exemplar.rb:13]:
```suggestion
    scores.delete(name)
```

Because at some point, the attribute reader should be defined:

```ruby
attr_reader :scores
```

This is preferred, because if we mistype, as a developer, the name of the instance variable, we get errors relating to `nil`, rather than a more helpful exception.

File: exercises/concept/high-score/.meta/exemplar.rb [hunk 3]
```diff
@@ -1,31 +1,31 @@
...
   end
 
-  def self.reset_player_score(score_board, name)
-    score_board[name] = INITIAL_SCORE
-    score_board
+  def reset_player_score(name)
+    @scores[name] = INITIAL_SCORE
```
Comment (kotp) [inline exercises/concept/high-score/.meta/exemplar.rb:17]: If we set up the attribute accessor this can be `scores[name] = INITIAL_SCORE`.

**Review by kotp — CHANGES_REQUESTED** [review]

File: exercises/concept/high-score/.meta/exemplar.rb (hunk 2 above)
Comment (iHiD) [reply exercises/concept/high-score/.meta/exemplar.rb:13]: I don't think we've not taught `attr_reader` at this stage in the track, so we can't use it here.

Comment (iHiD) [reply exercises/concept/high-score/.meta/exemplar.rb:13]: (We can, of course, choose to teach it in this exercise or before this, but that's a decision that needs to be made in order for us to be allowed to use it here)

Comment (kotp) [reply exercises/concept/high-score/.meta/exemplar.rb:13]: I saw "exemplar", so was not aware that it would go to the student.

Comment (iHiD) [reply exercises/concept/high-score/.meta/exemplar.rb:13]: Yeah. See the [spec for examplar files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-exemplar-implementation). Exemplars are the solutions we want the student to aim for at this point in their learning.

File: exercises/concept/high-score/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+To make a new high score map, define the `HighScore#initialize` function which doesn't take any arguments and initializes the new Scoreboard object instance. It should take an optional argument, with the default value of an empty hash.
+
+```ruby
+HighScore.new
+```
+
+## 2. Add players to the high score map
+
+To add a player to the high score map, define `HighScore.add_player/3`, which is a function which takes 3 arguments:
```
Comment (angelikatyborska) [inline exercises/concept/high-score/.docs/instructions.md:15]: I see some Elixir leftovers:
```suggestion
To add a player to the high score map, define `HighScore#add_player`, which is a function which takes 2 arguments:
```

File: exercises/concept/high-score/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
+# Instructions
+
+In this exercise, you're implementing a way to keep track of the high scores for the most popular game in your local arcade hall.
+
+## 1. Define a new high score map
+
+To make a new high score map, define the `HighScore#initialize` function which doesn't take any arguments and initializes the new Scoreboard object instance. It should take an optional argument, with the default value of an empty hash.
```
Comment (angelikatyborska) [inline exercises/concept/high-score/.docs/instructions.md:7]: The instructions use `HighScore` as the class name, but the example solution uses `ScoreBoard`. Also, should we be saying "methods" instead of "functions"?

File: exercises/concept/high-score/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+To get a player's score, define `HighScore#player_score` which takes one argument:
+
+- The first argument is the name of the player, whose score should be returned.
+
+```ruby
+score_map = HighScore.new
+score_map.add_player("Dave Thomas", 2_374)
+score_map.add_player("José Valim", 486_373)
+score_map.get_players
+# => ["Dave Thomas", "José Valim"]
```
Comment (angelikatyborska) [inline exercises/concept/high-score/.docs/instructions.md:78]: Wrong method:
```suggestion
score_map.player_score(player_score)
# => 2374
```

File: exercises/concept/high-score/high_score.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def update_player_score(name, score)
+    raise NotImplementedError, 'Please implement the Scoreboard#update_player_score method'
+  end
+
+  def player_score(name)
+    raise NotImplementedError, 'Please implement the Scoreboard#player_score method'
+  end
+end
```
Comment (angelikatyborska) [inline exercises/concept/high-score/high_score.rb:29]: New approach to stubs
```suggestion
class Scoreboard
  def initialize(initial_scores)
    raise 'Please implement the Scoreboard#initialize method'
  end

  def players
    raise 'Please implement the Scoreboard#players method'
  end

  def add_player(name, initial_score)
    raise 'Please implement the Scoreboard#add_player method'
  end

  def remove_player(name)
    raise 'Please implement the Scoreboard#remove_player method'
  end

  def reset_player_score(name)
    raise 'Please implement the Scoreboard#reset_player_score method'
  end

  def update_player_score(name, score)
    raise 'Please implement the Scoreboard#update_player_score method'
  end

  def player_score(name)
    raise 'Please implement the Scoreboard#player_score method'
  end
end
```

File: exercises/concept/high-score/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+## 7. Get a list of players
+
+To get a list of players, define `HighScore#players`.
+
+```ruby
+score_map = HighScore.new
+score_map.add_player("Dave Thomas", 2_374)
+score_map.add_player("José Valim", 486_373)
+score_map.get_players
```
Comment (angelikatyborska) [inline exercises/concept/high-score/.docs/instructions.md:89]: Method name mismatch:
```suggestion
score_map.players
```

File: exercises/concept/high-score/high_score_test.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
+require 'minitest/autorun'
+require_relative '.meta/exemplar'
+
+class HighScoreTest < Minitest::Test
+  def test_blank
+    expected = {}
+    scoreboard = Scoreboard.new
+    assert_equal [], scoreboard.players
```
Comment (angelikatyborska) [inline exercises/concept/high-score/high_score_test.rb:8]: The `players` method is being tested in the very first test case and in the very last test case. In the stub, it's defined at the beginning of the file, yet in the instructions it's the very last task to implement it. This might annoy the students who expect the tasks and tests to be in roughly the same order.

File: yarn.lock [hunk 10]
```diff
@@ -4,19 +4,19 @@
 
 "@babel/code-frame@^7.0.0":
   version "7.12.13"
-  resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.12.13.tgz#dcfc826beef65e75c50e21d3837d7d95798dd658"
+  resolved "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.12.13.tgz"
```
Comment (angelikatyborska) [inline yarn.lock:7]: Is this what happens when you run `npm install` instead of `yarn install`? I would guess this change should be excluded from the PR.

**Review by angelikatyborska — COMMENTED** [review]
I know this is not my track but I'm trying to help out with the content :) I hope my guest review is useful.

File: exercises/concept/high-score/.docs/instructions.md (hunk 5 above)
Comment (kotp) [reply exercises/concept/high-score/.docs/instructions.md:7]: Yes, we should use the name "method" rather than "function".

File: yarn.lock (hunk 10 above)
Comment (neenjaw) [reply yarn.lock:7]: 🤷🏻 I would assume that npm install wouldn't touch a yarn.lock at all

File: exercises/concept/high-score/.docs/instructions.md (hunk 6 above)
Comment (neenjaw) [reply exercises/concept/high-score/.docs/instructions.md:78]:
```suggestion
score_map.player_score("Dave Thomas")
# => 2374
```

**Review by kotp — DISMISSED** [review]

**Comment by kotp** [comment]
Still conflicted though.  But WIP so probably doesn't matter at this point.

File: exercises/concept/high-score/high_score_test.rb (hunk 9 above)
Comment (neenjaw) [reply exercises/concept/high-score/high_score_test.rb]: I have adapted the exercise/tests to not depend on that anymore :)

**Comment by neenjaw** [comment]
Screenshot from 2021-08-31 23-01-40

I think this happened automatically when I pushed the most recent commit.  Not sure if there was anything lost in the dismissal.

Yea, I did not address the config at all yet

**Comment by kotp** [comment]
> Screenshot from 2021-08-31 23-01-40
> 
> I think this happened automatically when I pushed the most recent commit. Not sure if there was anything lost in the dismissal.

It did, and nothing was lost.  Still WIP so the approval wasn't really expected to stick.

**Comment by kytrinyx** [comment]
It's been a year since the last movement on this PR.

Does someone who was involved at the time have an idea of what it would take to get this over the finish line?

**Comment by kotp** [comment]
This got bumped today, for some reason.  But I looked for a mention of removing `yarn.lock`, or a reason it might be here, and did not see it mentioned.

File: exercises/concept/high-score/.docs/hints.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
+# Hints
+
+If you get stuck, review the standard library documentation about:
+
+- [Classes][rubydoc-classes]
+- [The `Hash` class][rubydoc-hash]
+
+## 1. Define a new high score map
+
+- When creating a new instance of a class, the `initialize` method is invoked behind the scenes. Create a new [instance variable][ruby-user-guide-instance-variables] to keep the state of the class.
```
Comment (kotp) [inline exercises/concept/high-score/.docs/hints.md:10]: Single sentence per line, as per Exercism's [markdown specifications].

[markdown specifications]: https://exercism.org/docs/building/markdown/markdown

**Review by kotp — COMMENTED** [review]
Still, draft, so a comment.
